### PR TITLE
simplify transformer notebook

### PR DESCRIPTION
**Pull request by szha**
## Description ##
simplify transformer notebook. 

### Essentials ###
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented

### Changes ###
- [x] reduce notebook length by keeping only the inference part. the training part is commented out and should be added as a separate notebook. (that part of code is saved [here](https://gist.github.com/szha/ffde8760d2ed296c5022dfed6a04fbb9))
- [x] add the explanation to the usage of classes to help understand the content.

### CODE REVIEW

**Comment by mli** [comment]
Job PR-400/7 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-400/7/index.html

**Comment by mli** [comment]
Job PR-400/14 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-400/14/index.html

**Comment by mli** [comment]
Job PR-400/15 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-400/15/index.html

**Comment by mli** [comment]
Job PR-400/16 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-400/16/index.html

File: docs/examples/machine_translation/transformer.md [hunk 1]
```diff
@@ -57,71 +55,66 @@ The Transformer model architecture is shown as below:
...
 
 ```{.python .input}
 import dataprocessor
 
 print(dataprocessor.TrainValDataTransform.__doc__)
-```
 
-```{.python .input}
-wmt_transform_fn = dataprocessor.TrainValDataTransform(wmt_src_vocab, wmt_tgt_vocab, -1, -1)
+wmt_transform_fn = dataprocessor.TrainValDataTransform(wmt_src_vocab, wmt_tgt_vocab)
```
[the remark this answers is not in the document]
Comment (szha) [reply docs/examples/machine_translation/transformer.md]: Done.

File: docs/examples/machine_translation/transformer.md [hunk 2]
```diff
@@ -30,13 +28,13 @@ mx.random.seed(10000)
...
+## Use the Pretrained Transformer model
 
-In this subsection, we first load the SOTA Transformer model in GluonNLP model zoo; and secondly we load the full WMT 2014 English-German test dataset; and finally evaluate the model.
+Next, we load the Transformer model in GluonNLP model zoo and use the full newstest2014 segment of WMT 2014 English-German test dataset, and evaluate the model on it.
 
-### Get the SOTA Transformer
+### Get the Transformer
 
-Next, we load the pretrained SOTA Transformer using the model API in GluonNLP. In this way, we can easily get access to the SOTA machine translation model and use it in your own application.
+We load the pretrained Transformer using the model API in GluonNLP, which returns the source and target vocabulary along with the model.
```
[the remark this answers is not in the document]
Comment (szha) [reply docs/examples/machine_translation/transformer.md]: Actually, I'm confused about this too... is this a bug?

File: docs/examples/machine_translation/transformer.md [hunk 3]
```diff
@@ -132,22 +125,19 @@ wmt_test_batchify_fn = nlp.data.batchify.Tuple(
     nlp.data.batchify.Stack())
```
[the remark this answers is not in the document]
Comment (szha) [reply docs/examples/machine_translation/transformer.md]: Done.

**Comment by mli** [comment]
Job PR-400/18 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-400/18/index.html

File: docs/examples/machine_translation/transformer.md [hunk 4]
```diff
@@ -49,79 +47,75 @@ wmt_transformer_model, wmt_src_vocab, wmt_tgt_vocab = \
...
 ```{.python .input}
 import hyperparameters as hparams
 
 wmt_data_test = nlp.data.WMT2014BPE('newstest2014',
                                     src_lang=hparams.src_lang,
-                                    tgt_lang=hparams.tgt_lang,
-                                    full=False)
+                                    tgt_lang=hparams.tgt_lang)
 print('Source language %s, Target language %s' % (hparams.src_lang, hparams.tgt_lang))
+print('Sample: "{}"'.format(wmt_data_test[0]))
```
Comment (cgraywang) [inline docs/examples/machine_translation/transformer.md]: change to `print('Sample BPE-based Text: "{}"'.format(wmt_data_test[0]))` could be more user-friendly?

File: docs/examples/machine_translation/transformer.md [hunk 5]
```diff
@@ -49,79 +47,75 @@ wmt_transformer_model, wmt_src_vocab, wmt_tgt_vocab = \
...
-
-```{.python .input}
 wmt_test_text = nlp.data.WMT2014('newstest2014',
                                  src_lang=hparams.src_lang,
-                                 tgt_lang=hparams.tgt_lang,
-                                 full=False)
-wmt_test_text[0]
-```
+                                 tgt_lang=hparams.tgt_lang)
+print('Sample text: "{}"'.format(wmt_test_text[0]))
```
Comment (cgraywang) [inline docs/examples/machine_translation/transformer.md]: `print('Sample Raw Text: "{}"'.format(wmt_test_text[0]))`

File: docs/examples/machine_translation/transformer.md [hunk 6]
```diff
@@ -49,79 +47,75 @@ wmt_transformer_model, wmt_src_vocab, wmt_tgt_vocab = \
...
-```{.python .input}
-wmt_transform_fn = dataprocessor.TrainValDataTransform(wmt_src_vocab, wmt_tgt_vocab, -1, -1)
+# wmt_transform_fn includes the four preprocessing steps mentioned above.
+wmt_transform_fn = dataprocessor.TrainValDataTransform(wmt_src_vocab, wmt_tgt_vocab)
 wmt_dataset_processed = wmt_data_test.transform(wmt_transform_fn, lazy=False)
 print(*wmt_dataset_processed[0], sep='\n')
-```
 
-### Create Sampler and DataLoader for WMT 2014 Dataset
+def get_length_index_fn():
```
Comment (cgraywang) [inline docs/examples/machine_translation/transformer.md]: We might want to hide this, we can integrate this `def get_length_index_fn()` in `dataprocessor`.

File: docs/examples/machine_translation/transformer.md [hunk 7]
```diff
@@ -132,22 +126,28 @@ wmt_test_batchify_fn = nlp.data.batchify.Tuple(
...
+* [Pad](../../api/data.batchify.rst#gluonnlp.data.batchify.Pad) takes the elements from all dataset items in a batch, and pad them according to the item of maximum length to form a padded matrix/tensor.
+* [Stack](../../api/data.batchify.rst#gluonnlp.data.batchify.Stack) simply stacks all elements in a batch, and requires all elements to be of the same length.
 
-```{.python .input}
-wmt_bucket_scheme = nlp.data.ExpWidthBucket(bucket_len_step=1.2)
-```
+
+We can then construct bucketing samplers, which generate batches by grouping sequences with similar lengths. Here, we use [FixedBucketSampler](../../api/data.rst#gluonnlp.data.FixedBucketSampler) with [ExpWidthBucket](../../api/data.rst#gluonnlp.data.ExpWidthBucket). With this setting, the sampler would select buckets following an approximately exponentially increasing interval of maximum bucket lengths.
```
Comment (cgraywang) [inline docs/examples/machine_translation/transformer.md]: We could explain the bucketsampler a little bit. Insert this sentence `FixedBucketSampler aims to assign each data sample to a fixed bucket based on its length.` before `With this setting, the sampler would select buckets following an approximately exponentially increasing interval of maximum bucket lengths.`

File: docs/examples/machine_translation/transformer.md [hunk 8]
```diff
@@ -228,7 +225,7 @@ import utils
 
 print('Translate the following English sentence into German:')
 
-sample_src_seq = 'We love each other'
+sample_src_seq = 'We love language.'
```
Comment (cgraywang) [inline docs/examples/machine_translation/transformer.md]: 😄

File: docs/examples/machine_translation/transformer.md [hunk 9]
```diff
@@ -243,248 +240,11 @@ print('The German translation is:')
 print(sample_tgt_seq)
 ```
 
-## Train Your Own Transformer
-
-In this subsection, we will go though the whole process about loading translation dataset in a more unified way, and create data sampler and loader, as well as define the Transformer model, finally writing training script to train the model yourself.
+## Next
```
Comment (cgraywang) [inline docs/examples/machine_translation/transformer.md]: Since the Next here is a little bit duplicated with the Next button on the page, we can probably remove the section title.
